File: sdks/unity/Assets/AfricaPaymentsMcp/Scripts/Models/PaymentResponse.cs
using System;
using UnityEngine;

namespace AfricaPaymentsMcp.Models
{
    /// <summary>
    /// Response from a payment operation
    /// </summary>
    [Serializable]
    public class PaymentResponse
    {
        [SerializeField] private string transactionId;
        [SerializeField] private PaymentStatus status;
        [SerializeField] private string reference;
        [SerializeField] private double amount;
        [SerializeField] private string currency;
        [SerializeField] private string phoneNumber;
        [SerializeField] private string createdAt;
        [SerializeField] private string updatedAt;
        [SerializeField] private string receiptUrl;
        [SerializeField] private string failureReason;

        public string TransactionId 
        { 
            get => transactionId; 
            set => transactionId = value; 
        }
        
        public PaymentStatus Status 
        { 
            get => status; 
            set => status = value; 
        }
        
        public string Reference 
        { 
            get => reference; 
            set => reference = value; 
        }
        
        public double Amount 
        { 
            get => amount; 
            set => amount = value; 
        }
        
        public string Currency 
        { 
            get => currency; 
            set => currency = value; 
        }
        
        public string PhoneNumber 
        { 
            get => phoneNumber; 
            set => phoneNumber = value; 
        }
        
        public string CreatedAt 
        { 
            get => createdAt; 
            set => createdAt = value; 
        }
        
        public string UpdatedAt 
        { 
            get => updatedAt; 
            set => updatedAt = value; 
        }
        
        public string ReceiptUrl 
        { 
            get => receiptUrl; 
            set => receiptUrl = value; 
        }
        
        public string FailureReason 
        { 
            get => failureReason; 
            set => failureReason = value; 
        }

        /// <summary>
        /// Checks if payment was successful
        /// </summary>
        public bool IsSuccess => status == PaymentStatus.Success;

        /// <summary>
        /// Checks if payment is pending
        /// </summary>
        public bool IsPending => status == PaymentStatus.Pending;

        /// <summary>
        /// Checks if payment failed
        /// </summary>
        public bool IsFailed => status == PaymentStatus.Failed;

        /// <summary>
        /// Creates a PaymentResponse from JSON
        /// </summary>
        public static PaymentResponse FromJson(string json)
        {
            var dict = MiniJSON.Json.Deserialize(json) as System.Collections.Generic.Dictionary<string, object>;
            if (dict == null) return null;

            var response = new PaymentResponse
            {
                transactionId = GetString(dict, "transactionId"),
                reference = GetString(dict, "reference"),
                amount = GetDouble(dict, "amount"),
                currency = GetString(dict, "currency"),
                phoneNumber = GetString(dict, "phoneNumber"),
                createdAt = GetString(dict, "createdAt"),
                updatedAt = GetString(dict, "updatedAt"),
                receiptUrl = GetString(dict, "receiptUrl"),
                failureReason = GetString(dict, "failureReason")
            };

            if (dict.TryGetValue("status", out var statusObj) && statusObj is string statusStr)
            {
                response.status = statusStr.ToLower() switch
                {
                    "pending" => PaymentStatus.Pending,
                    "success" => PaymentStatus.Success,
                    "failed" => PaymentStatus.Failed,
                    "cancelled" => PaymentStatus.Cancelled,
                    _ => PaymentStatus.Pending
                };
            }

            return response;
        }

        private static string GetString(System.Collections.Generic.Dictionary<string, object> dict, string key)
        {
            return dict.TryGetValue(key, out var value) ? value?.ToString() : null;
        }

        private static double GetDouble(System.Collections.Generic.Dictionary<string, object> dict, string key)
        {
            if (dict.TryGetValue(key, out var value))
            {
                if (value is double d) return d;
                if (value is float f) return f;
                if (double.TryParse(value?.ToString(), out var result)) return result;
            }
            return 0;
        }
    }

    /// <summary>
    /// Transaction history response
    /// </summary>
    [Serializable]
    public class TransactionHistory
    {
        public PaymentResponse[] transactions;
        public int total;
        public bool hasMore;
    }
}

File: sdks/flutter/lib/src/models/payment_response.dart
import 'package:equatable/equatable.dart';
import 'package:json_annotation/json_annotation.dart';

part 'payment_response.g.dart';

/// Payment status
enum PaymentStatus {
  @JsonValue('pending')
  pending,
  @JsonValue('success')
  success,
  @JsonValue('failed')
  failed,
  @JsonValue('cancelled')
  cancelled,
}

/// Response from a payment operation
@JsonSerializable()
class PaymentResponse extends Equatable {
  /// Unique transaction ID
  final String transactionId;

  /// Current status
  final PaymentStatus status;

  /// Reference provided in the request
  final String reference;

  /// Payment amount
  final double amount;

  /// Currency code
  final String currency;

  /// Phone number used
  final String phoneNumber;

  /// Creation timestamp
  final DateTime createdAt;

  /// Last update timestamp
  final DateTime updatedAt;

  /// Optional receipt URL
  final String? receiptUrl;

  /// Failure reason (if status is failed)
  final String? failureReason;

  const PaymentResponse({
    required this.transactionId,
    required this.status,
    required this.reference,
    required this.amount,
    required this.currency,
    required this.phoneNumber,
    required this.createdAt,
    required this.updatedAt,
    this.receiptUrl,
    this.failureReason,
  });

  factory PaymentResponse.fromJson(Map<String, dynamic> json) =>
      _$PaymentResponseFromJson(json);

  Map<String, dynamic> toJson() => _$PaymentResponseToJson(this);

  /// Check if payment was successful
  bool get isSuccess => status == PaymentStatus.success;

  /// Check if payment is pending
  bool get isPending => status == PaymentStatus.pending;

  /// Check if payment failed
  bool get isFailed => status == PaymentStatus.failed;

  @override
  List<Object?> get props => [
        transactionId,
        status,
        reference,
        amount,
        currency,
        phoneNumber,
        createdAt,
        updatedAt,
        receiptUrl,
        failureReason,
      ];
}

/// Transaction history response
@JsonSerializable()
class TransactionHistory extends Equatable {
  /// List of transactions
  final List<PaymentResponse> transactions;

  /// Total count
  final int total;

  /// Whether there are more results
  final bool hasMore;

  const TransactionHistory({
    required this.transactions,
    required this.total,
    required this.hasMore,
  });

  factory TransactionHistory.fromJson(Map<String, dynamic> json) =>
      _$TransactionHistoryFromJson(json);

  Map<String, dynamic> toJson() => _$TransactionHistoryToJson(this);

  @override
  List<Object?> get props => [transactions, total, hasMore];
}
